src/Workspaces/Remote/Core/RemoteCallback.cs

```diff
@@ -87,23 +87,75 @@ public async ValueTask<TResult> InvokeAsync<TResult>(Func<T, CancellationToken,
         public async ValueTask<TResult> InvokeAsync<TResult>(
             Func<T, PipeWriter, CancellationToken, ValueTask> invocation,
             Func<PipeReader, CancellationToken, ValueTask<TResult>> reader,
             CancellationToken cancellationToken)
         {
             try
             {
-                return await BrokeredServiceConnection<T>.InvokeStreamingServiceAsync(_callback, invocation, reader, cancellationToken).ConfigureAwait(false);
+                return await InvokeWorkerAsync(_callback).ConfigureAwait(false);
             }
             catch (Exception exception) when (ReportUnexpectedException(exception, cancellationToken))
             {
                 throw OnUnexpectedException(exception, cancellationToken);
             }
+
+            async ValueTask<TResult> InvokeWorkerAsync(T service)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var pipe = new Pipe();
+
+                // Kick off the work to do the writing to the pipe in a fire-and-forget fashion.  It will start hot and
+                // will be able to do work as the reading side attempts to pull in the data it is writing.
+
+                _ = WriteAsync(service, pipe.Writer);
+
+                try
+                {
+                    return await reader(pipe.Reader, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    // ensure we always complete the reader so the pipe can clean up all its resources. in the case of
+                    // an exception, attempt to complete the reader with that as well as that will tear down the writer
+                    // allowing it to stop writing and allowing the pipe to be cleaned up.
+                    await pipe.Reader.CompleteAsync(e).ConfigureAwait(false);
+                    throw;
+                }
+                finally
+                {
+                    // ensure we always complete the reader so the pipe can clean up all its resources.
+                    await pipe.Reader.CompleteAsync().ConfigureAwait(false);
+                }
+            }
+
+            async Task WriteAsync(T service, PipeWriter writer)
+            {
+                try
+                {
+                    await invocation(service, writer, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    // Ensure that the writer is complete if an exception is thrown. This intentionally swallows the
+                    // exception on this side, knowing it will actually be thrown on the reading side.
+                    await writer.CompleteAsync(e).ConfigureAwait(false);
+                }
+                finally
+                {
+                    await writer.CompleteAsync(exception: null).ConfigureAwait(false);
+                }
+            }
```

Comment:
simple wrapper around the async `invocation` callback that simply ensures we always complete the writer (optionally with a failure exception).